src/allocation_engine/interval_tree.rs

```diff
@@ -388,8 +390,39 @@ impl InnerNode {
                 }
             },
         }
-        Ok(self.update_node())
+        Ok(self.update_height_and_balance())
     }
+
+    /// Update the state of an old node. This method should be used when we
+    /// want to allocate a free node.
+    #[allow(dead_code)]
+    pub(crate) fn update(&mut self, key: &Range, node_state: NodeState) -> Result<()> {
```

Comment:
You are mentioning in the commit message that the only valid transition is from free to allocated. In this case, we can make this function impossible to be used wrong, and as a consequence simplify the code. This will eliminate some transitions, making the code less prone to error.

A few changes would be needed:
- remove the node_state parameter, we're only going to set the `node_state` to Allocated
- rename the function such that it suggests the transition is from free to allocated
- check that the node we want to update is Free

Reply:
I have the same idea with this method :) Previously the `update` here will also update the data stored in the Node, but as we remove the design that every node stores some data, this part should also have some slight changes.

Reply:
This still has the `node_state` parameter. Can you either explain why it's needed or remove it?